compiler/src/iree/compiler/Codegen/Common/LinearizeMemRefs.cpp

```diff
@@ -0,0 +1,455 @@
+      return dims;
+    } else {
+      return failure();
+    }
+  }
+}
+
+template <typename OpTy>
+struct LinearizeMemrefAlloc : public OpRewritePattern<OpTy> {
+  LinearizeMemrefAlloc(MLIRContext *context)
+      : OpRewritePattern<OpTy>(context) {}
+
+  LogicalResult matchAndRewrite(OpTy allocOp,
+                                PatternRewriter &rewriter) const override {
+    static_assert(std::is_same<OpTy, memref::AllocOp>() ||
+                      std::is_same<OpTy, memref::AllocaOp>(),
+                  "expected only memref::AllocOp or memref::AllocaOp");
+    Location loc = allocOp->getLoc();
+    MemRefType currentTypeOfSourceMemref =
+        dyn_cast<MemRefType>(allocOp.getMemref().getType());
+    if (currentTypeOfSourceMemref.getRank() < 2)
+      return success();
+    FailureOr<MemRefType> maybeNewTypeOfSourceMemref =
+        getLinearizedTypeFromSourceType(currentTypeOfSourceMemref);
+    if (failed(maybeNewTypeOfSourceMemref))
+      return failure();
+    MemRefType newTypeOfSourceMemref = *maybeNewTypeOfSourceMemref;
+
+    SmallVector<Value> dynamicLinearizedSize;
+    if (!newTypeOfSourceMemref.hasStaticShape()) {
+      SmallVector<Value> basis = getDimValues(
+          loc, rewriter, currentTypeOfSourceMemref, allocOp.getDynamicSizes());
+      SmallVector<Value> multiIndices(
+          basis.size(), rewriter.create<arith::ConstantIndexOp>(loc, 0));
+      multiIndices[0] = basis[0];
+      Value linearizedSizes = rewriter.create<affine::AffineLinearizeIndexOp>(
+          loc, multiIndices, basis, true);
+      dynamicLinearizedSize.push_back(linearizedSizes);
+    }
+
+    Value linearizedOp = rewriter.create<OpTy>(
+        loc, newTypeOfSourceMemref, dynamicLinearizedSize,
+        allocOp.getSymbolOperands(), allocOp.getAlignmentAttr());
+    SmallVector<int64_t, 2> indices(currentTypeOfSourceMemref.getRank());
+    std::iota(indices.begin(), indices.end(), 0);
+    Value delinearizedOp = rewriter.create<memref::ExpandShapeOp>(
+        loc, currentTypeOfSourceMemref, linearizedOp, ArrayRef({indices}),
+        allocOp.getMixedSizes());
+    rewriter.replaceOp(allocOp, delinearizedOp);
+    return success();
+  }
+};
+
+struct LinearizeMemrefLoad : public OpRewritePattern<memref::LoadOp> {
+  using OpRewritePattern<memref::LoadOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(memref::LoadOp loadOp,
+                                PatternRewriter &rewriter) const override {
+    Location loc = loadOp->getLoc();
+    MemRefType currentTypeOfSourceMemref = loadOp.getMemRefType();
+    if (currentTypeOfSourceMemref.getRank() < 2 &&
+        loadOp.getIndices().size() < 2)
+      return success();
+    FailureOr<MemRefType> maybeNewTypeOfSourceMemref =
+        getLinearizedTypeFromSourceType(currentTypeOfSourceMemref);
+    if (failed(maybeNewTypeOfSourceMemref))
+      return failure();
+    MemRefType newTypeOfSourceMemref = *maybeNewTypeOfSourceMemref;
+
+    FailureOr<SmallVector<Value>> basis =
```

Comment:
Yeah, this should be `OpFoldResult` since it's a linearize  basis

Reply:
Done.